### DISCUSSION TITLE
What is the optimal reward system for SAP?

### DISCUSSION THREAD

**Discussion by andreped** [andreped/super-ml-pets#20] [Ideas]
I think the reward system should capture these aspect (some of which are likely not covered yet):
- Reward the AI if it wins, penalize if it loses.
- Reward the AI if it uses ALL the money (better to roll away and freeze stuff, than skipping).
- For smarter learning it is likely smart to reward minimizing number of actions between battles (to some extent) at it should help with reducing random shuffling which is not relevant (but not be too strict here, otherwise it may not learn smarter strategies).
- Improve opponent generation using curriculum learning (easy opponents at the start, harder opponents over time).
- Reward win streaks? In a single game for individual battles? Might be that it hinders the AI to learn strategies where it is smarter to lose some games in order to build the best time (e.g., lots of snails in store).

If you find some ideas that are missing, please, feel free to share different ideas :) The ones listed above are those that I fundamentally feel would improve performance, quite rapidly so.

**Comment by GoldExplosion** [comment]
> Reward the AI if it wins, penalize if it loses. 

and 
> Reward win streaks

hinders the AI to learn strategies where it is smarter to lose some games in order to build the best time.
It might be better to reward the AI based on number of total wins after it loses all of its lives or reaches 10 wins.

**Reply by andreped** [reply]
I agree. I just thought it was relevant to start the discussion. But as discussed, having a proper reward system is extremely important, and I believe that is the next step in improving performance. Currently, the model is performing well in training in the simulated environment, but as the simulated environment does not well mimic battles against real opponents, we are struggling to get similar/improved performance.

It would be great if you could experiment with the reward system. The easiest way of doing it is by forking `sapai-gym` and adding the custom reward system there. Note that there is an ongoing PR (draft) to support alternative reward functions, but seems to be not active anymore:
https://github.com/alexdriedger/sapai-gym/pull/3

Might be a good place to start :) When testing, be sure to update your fork of `super-ml-pets` to use your own fork of `sapai-gym` btw.
